[blend2d] Update to 2022-02-19

ZeeWanderer opened the pull request:
- Updated `blend2d` to `2022-02-19` and embedded `asmjit` library to `2022-02-20`.
 - Added new **default** feature [`futex`](https://github.com/blend2d/blend2d/blob/master/src/blend2d/threading/futex_p.h).
 - Used `vcpkg-cmake` as required
 - Added `license` field
 - Resolved conflicts with #22741

LilyWangLL commented:
Thanks for your PR. `osgearth` build failed with the following error. It depends on `blend2d`.
```
D:\buildtrees\osgearth\src\06c3d361c0-e3aed5a168.clean\src\osgEarth\FeatureRasterizer.cpp
D:\buildtrees\osgearth\src\06c3d361c0-e3aed5a168.clean\src\osgEarth\FeatureRasterizer.cpp(281): error C2664: 'BLResult BLContext::setStrokeCaps(BLStrokeCap) noexcept': cannot convert argument 1 from 'uint32_t' to 'BLStrokeCap'
D:\buildtrees\osgearth\src\06c3d361c0-e3aed5a168.clean\src\osgEarth\FeatureRasterizer.cpp(281): note: Conversion to enumeration type requires an explicit cast (static_cast, C-style cast or function-style cast)
D:\installed\x64-windows\include\blend2d/context.h(1186): note: see declaration of 'BLContext::setStrokeCaps'
D:\buildtrees\osgearth\src\06c3d361c0-e3aed5a168.clean\src\osgEarth\FeatureRasterizer.cpp(282): error C2664: 'BLResult BLContext::setStrokeJoin(BLStrokeJoin) noexcept': cannot convert argument 1 from 'uint32_t' to 'BLStrokeJoin'
D:\buildtrees\osgearth\src\06c3d361c0-e3aed5a168.clean\src\osgEarth\FeatureRasterizer.cpp(282): note: Conversion to enumeration type requires an explicit cast (static_cast, C-style cast or function-style cast)
D:\installed\x64-windows\include\blend2d/context.h(1178): note: see declaration of 'BLContext::setStrokeJoin'
D:\buildtrees\osgearth\src\06c3d361c0-e3aed5a168.clean\src\osgEarth\FeatureRasterizer.cpp(940): warning C4018: '<': signed/unsigned mismatch
D:\buildtrees\osgearth\src\06c3d361c0-e3aed5a168.clean\src\osgEarth\FeatureRasterizer.cpp(949): warning C4018: '<': signed/unsigned mismatch
```

ZeeWanderer commented:
blend2d is in beta, breaking changes are a given. For now, i suppose osgearth port can depend on blend2d version it supports via override until upstream is fixed. Breaking commit is [be3c0583](https://github.com/blend2d/blend2d/commit/be3c0583427f706b97b9c3fdbd8e637c36bee6d7), `BLStrokeCap ` enum used in mentioned functions instead of `uint32_t`.

LilyWangLL commented:
> 

Thanks for your reply. Is there a PR or issue on upstream to resolve this bug? If there is, could you please link it to this PR?

ZeeWanderer commented:
Added an issue.

ZeeWanderer commented:
Considering last `osgearth` version was released at aug 9 and new version is TBD (only 45% done according to Milestones) i think it would be better to modify `osgearth` port and go ahead with this pr. Unfortunately, i can't quite get versioning deps to work. Tried overriding and setting `builtin-baseline` commit to one containing only last supported `blen2d` version.

ZeeWanderer commented:
New description:
 - Updated `blend2d` to `2022-02-19` and embedded `asmjit` library to `2022-02-20`.
 - Added new **default** feature [`futex`](https://github.com/blend2d/blend2d/blob/master/src/blend2d/threading/futex_p.h).
 - Used `vcpkg-cmake` as required
 - Added `license` field
 - Resolved conflicts with #22741

LilyWangLL commented:
`osgearth` depends on `blend2d`, it build failed with the error on pipeline:
```
C:\PROGRA~2\MICROS~3\2019\ENTERP~1\VC\Tools\MSVC\1429~1.301\bin\Hostx64\x64\cl.exe   /TP -DGEOS_INLINE -DH5_BUILT_AS_DYNAMIC_LIB -DOSGEARTH_ENABLE_IMGUI -DOSGEARTH_HAVE_BLEND2D -DOSGEARTH_HAVE_GEOS -DOSGEARTH_HAVE_MBTILES -DOSGEARTH_HAVE_MVT -DOSGEARTH_HAVE_SQLITE3 -DOSGEARTH_LIBRARY -DPROTOBUF_USE_DLLS -DSQLITE_API=__declspec(dllimport) -DSQLITE_ENABLE_RTREE -DSQLITE_OS_WIN=1 -DTIXML_USE_STL -DUSE_UNSTABLE_GEOS_CPP_API -D_CRT_SECURE_NO_DEPRECATE -D_SCL_SECURE_NO_WARNINGS -D__STDC_FORMAT_MACROS -DosgEarth_EXPORTS -ID:\buildtrees\osgearth\x64-windows-dbg\build_include -ID:\buildtrees\osgearth\src\06c3d361c0-e3aed5a168.clean\src -ID:\buildtrees\osgearth\x64-windows-dbg\src\osgEarth -ID:\installed\x64-windows\include -ID:\installed\x64-windows\include\geos -ID:\installed\x64-windows\include\server -ID:\installed\x64-windows\include\libxml2 /nologo /DWIN32 /D_WINDOWS /W3 /utf-8 /GR /EHsc /MP  /D_DEBUG /MDd /Z7 /Ob0 /Od /RTC1  /fp:precise /wd4996 /showIncludes /Fosrc\osgEarth\CMakeFiles\osgEarth.dir\FeatureRasterizer.cpp.obj /Fdsrc\osgEarth\CMakeFiles\osgEarth.dir\ /FS -c D:\buildtrees\osgearth\src\06c3d361c0-e3aed5a168.clean\src\osgEarth\FeatureRasterizer.cpp
D:\buildtrees\osgearth\src\06c3d361c0-e3aed5a168.clean\src\osgEarth\FeatureRasterizer.cpp(281): error C2664: 'BLResult BLContext::setStrokeCaps(BLStrokeCap) noexcept': cannot convert argument 1 from 'uint32_t' to 'BLStrokeCap'
D:\buildtrees\osgearth\src\06c3d361c0-e3aed5a168.clean\src\osgEarth\FeatureRasterizer.cpp(281): note: Conversion to enumeration type requires an explicit cast (static_cast, C-style cast or function-style cast)
D:\installed\x64-windows\include\blend2d/context.h(1228): note: see declaration of 'BLContext::setStrokeCaps'
D:\buildtrees\osgearth\src\06c3d361c0-e3aed5a168.clean\src\osgEarth\FeatureRasterizer.cpp(282): error C2664: 'BLResult BLContext::setStrokeJoin(BLStrokeJoin) noexcept': cannot convert argument 1 from 'uint32_t' to 'BLStrokeJoin'
D:\buildtrees\osgearth\src\06c3d361c0-e3aed5a168.clean\src\osgEarth\FeatureRasterizer.cpp(282): note: Conversion to enumeration type requires an explicit cast (static_cast, C-style cast or function-style cast)
```

ZeeWanderer commented:
@LilyWangLL `blend2d` is beta, breaking changes are a given. There already is an issue for this on `osgEarth` i mentioned before but next version and therefore next port is TBD. The only sane solution i see is to use versioning and change `osgEarth` port to depend on earlier version of `blend2d`. I tried to do that but did not manage to get it working.

Alternatively, the patch for `osgEarth` should be simple 2 lines of change.

Which would be the preferable solution?

LilyWangLL commented:
> @LilyWangLL `blend2d` is beta, breaking changes are a given. There already is an issue for this on `osgEarth` i mentioned before but next version and therefore next port update is TBD. The only sane solution i see is to use versioning and change `osgEarth` port to depend on earlier version of `blend2d`. I tried to do that but did not manage to get it working.
> 
> Alternatively, the patch for `osgEarth` should be simple, 2 lines of change.
> 
> Which would be the preferable solution?

Can you submit a PR on https://github.com/gwaldron/osgearth to fix this error? Until upsteam PR merged, add a patch on VCPKG.

ZeeWanderer commented:
Sorry for the long wait, things are kind of hectic now. Made port fix for `osgearth`.

ZeeWanderer committed:
[blend2d] Update to 2022-02-19

ZeeWanderer committed:
[blend2d] update versions

strega-nil-ms commented:
Thanks!
